File: hs/cgen_ex/c/Dog.cpp
#define CGEN_OUTPUT_INTERN
#include "Dog.h"
void Animals_Dog_delete(Dog* this_ptr)
{
    delete this_ptr;
}

Dog* Animals_Dog_new()
{
    return new Dog();
}

void Animals_Dog_make_sound(Dog* this_ptr)
{
    this_ptr->make_sound();
}

File: hs/cgen_ex/hs/Dog.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module Dog(
dog_with, 
dog_delete, 
dog_new, 
dog_make_sound
)

where

import Types
import Control.Monad

import Foreign
import Foreign.C.String
import Foreign.C.Types

dog_with :: (Dog -> IO a) -> IO a
dog_with  f = do
    obj <- dog_new 
    res <- f obj
    dog_delete obj
    return res

foreign import ccall "Dog.h Animals_Dog_delete" c_dog_delete :: Dog -> IO ()
dog_delete :: Dog -> IO ()
dog_delete p1 =   c_dog_delete p1

foreign import ccall "Dog.h Animals_Dog_new" c_dog_new :: IO Dog
dog_new :: IO Dog
dog_new  =   c_dog_new 

foreign import ccall "Dog.h Animals_Dog_make_sound" c_dog_make_sound :: Dog -> IO ()
dog_make_sound :: Dog -> IO ()
dog_make_sound p1 =   c_dog_make_sound p1

